src/pal/src/thread/process.cpp

```diff
@@ -1758,6 +1770,236 @@ OpenProcess(
+    if (listHead != NULL)
+    {
+        for (ProcessModules *entry = listHead; entry != NULL; entry = entry->Next)
+        {
+            if ((HMODULE)entry->BaseAddress == hModule)
+            {
+                // Convert CHAR string into WCHAR string
+                result = MultiByteToWideChar(CP_ACP, 0, entry->Name, -1, lpFilename, nSize);
+                break;
+            }
+        }
+    }
+
+    DestroyProcessModules(listHead);
+
+    return result;
+}
+
+/*++
+Function:
+  CreateProcessModules
+
+Abstract
+  Returns a process's module list
+
+Return
+  ProcessModules * list
+
+--*/
+ProcessModules *
+CreateProcessModules(
+    IN HANDLE hProcess,
+    OUT LPDWORD lpCount)
+{
+    DWORD dwProcessId = PROCGetProcessIDFromHandle(hProcess);
+    if (dwProcessId == 0)
+    {
+        SetLastError(ERROR_INVALID_HANDLE);
+        return NULL;
+    }
+
+#ifdef HAVE_PROCFS_CTL 
+    // Here we read /proc/<pid>/maps file in order to parse it and figure out what it says 
+    // about a library we are looking for. This file looks something like this:
+    //
+    // [address]      [perms] [offset] [dev] [inode]     [pathname] - HEADER is not preset in an actual file
+    //
+    // 35b1800000-35b1820000 r-xp 00000000 08:02 135522  /usr/lib64/ld-2.15.so
+    // 35b1a1f000-35b1a20000 r--p 0001f000 08:02 135522  /usr/lib64/ld-2.15.so
+    // 35b1a20000-35b1a21000 rw-p 00020000 08:02 135522  /usr/lib64/ld-2.15.so
+    // 35b1a21000-35b1a22000 rw-p 00000000 00:00 0       [heap]
+    // 35b1c00000-35b1dac000 r-xp 00000000 08:02 135870  /usr/lib64/libc-2.15.so
+    // 35b1dac000-35b1fac000 ---p 001ac000 08:02 135870  /usr/lib64/libc-2.15.so
+    // 35b1fac000-35b1fb0000 r--p 001ac000 08:02 135870  /usr/lib64/libc-2.15.so
+    // 35b1fb0000-35b1fb2000 rw-p 001b0000 08:02 135870  /usr/lib64/libc-2.15.so
+
+    // Making something like: /proc/123/maps
+    char mapFileName[100]; 
+    int chars = snprintf(mapFileName, sizeof(mapFileName), "/proc/%d/maps", dwProcessId);
+    _ASSERTE(chars > 0 && chars <= sizeof(mapFileName));
+
+    FILE *mapsFile = fopen(mapFileName, "r");
+    if (mapsFile == NULL) 
+    {
+        SetLastError(ERROR_INVALID_HANDLE);
+        return NULL;
+    }
+
+    LockModuleList();
+
+    CPalThread* pThread = InternalGetCurrentThread();	
+    ProcessModules *listHead = NULL;
+    char *line = NULL;
+    size_t lineLen = 0;
+    int count = 0;
+    ssize_t read;
+
+    // Reading maps file line by line 
+    while ((read = getline(&line, &lineLen, mapsFile)) != -1) 
+    {
+        void *startAddress, *endAddress, *offset;
+        int devHi, devLo, inode;
+        char moduleName[PATH_MAX];
+
+        if (sscanf(line, "%p-%p %*[-rwxsp] %p %x:%x %d %s\n", &startAddress, &endAddress, &offset, &devHi, &devLo, &inode, moduleName) == 7)
+        {
+            if (inode != 0)
+            {
+                bool dup = false;
+                for (ProcessModules *entry = listHead; entry != NULL; entry = entry->Next)
```

Comment:
It is usually true except I did see a map like this:


7f18da0dc000-7f18da2db000 ---p 00025000 fc:00 1048889                    /lib/x86_64-linux-gnu/libtinfo.so.5.9
7f18da2db000-7f18da2df000 r--p 00024000 fc:00 1048889                    /lib/x86_64-linux-gnu/libtinfo.so.5.9
7f18da2df000-7f18da2e0000 rw-p 00028000 fc:00 1048889                    /lib/x86_64-linux-gnu/libtinfo.so.5.9
7f18da2e0000-7f18da304000 r-xp 00000000 fc:00 1050720                    /lib/x86_64-linux-gnu/ld-2.21.so
7f18da4e8000-7f18da4eb000 rw-p 00000000 00:00 0
7f18da4fa000-7f18da501000 r--s 00000000 fc:00 15999053                   /usr/lib/x86_64-linux-gnu/gconv/gconv-modules.cache
7f18da501000-7f18da503000 rw-p 00000000 00:00 0
7f18da503000-7f18da504000 r--p 00023000 fc:00 1050720                    /lib/x86_64-linux-gnu/ld-2.21.so
7f18da504000-7f18da505000 rw-p 00024000 fc:00 1050720                    /lib/x86_64-linux-gnu/ld-2.21.so
7f18da505000-7f18da506000 rw-p 00000000 00:00 0

And I guess I was just being thorough.

mikem

From: Eugene Zemtsov [mailto:notifications@github.com]
Sent: Monday, June 15, 2015 3:10 PM
To: dotnet/coreclr
Cc: Mike McLaughlin
Subject: Re: [coreclr] Fully implement dbgshim so it doesn't depend on LD_LIBRARY_PATH (issue #650) (#1141)


In src/pal/src/thread/process.cpp<https://github.com/dotnet/coreclr/pull/1141#discussion_r32472765>:

> +    int count = 0;

> +    ssize_t read;

> +

> +    // Reading maps file line by line

> +    while ((read = getline(&line, &lineLen, mapsFile)) != -1)

> +    {

> +        void *startAddress, *endAddress, *offset;

> +        int devHi, devLo, inode;

> +        char moduleName[PATH_MAX];

> +

> +        if (sscanf(line, "%p-%p %*[-rwxsp] %p %x:%x %d %s\n", &startAddress, &endAddress, &offset, &devHi, &devLo, &inode, moduleName) == 7)

> +        {

> +            if (inode != 0)

> +            {

> +                bool dup = false;

> +                for (ProcessModules *entry = listHead; entry != NULL; entry = entry->Next)

Is it really possible to have shared library mapped to non-contiguous blocks of memory? If not, we get rid of this loop and only check current list head.

—
Reply to this email directly or view it on GitHub<https://github.com/dotnet/coreclr/pull/1141/files#r32472765>.